httpbp/response.go

```diff
@@ -0,0 +1,135 @@
+package httpbp
+
+import (
+	"context"
+	"encoding/json"
+	"html/template"
+	"net/http"
+
+	httpgk "github.com/go-kit/kit/transport/http"
+)
+
+// ResponseCookies is an interface that your Response objects can implement in
+// order to have the httpbp.Encode methods automatically add cookies to the
+// response.
+type ResponseCookies interface {
+	// Return a list of all cookies to set on the response.
+	Cookies() []*http.Cookie
+
+	// Add a new cookie to set on the response.
+	AddCookie(*http.Cookie)
+}
+
+// ResponseHeaders is an interface that your Response objects can implement in
+// order to have the httpbp.Encode methods automatically add http headers to the
+// response.
+type ResponseHeaders interface {
+	// Headers returns the http.Header container that can be used to add headers
+	// to the response.
+	Headers() http.Header
+}
+
+// ErrorResponse is an interface that your Response objects can implement in
+// order to have the httpbp.Encode methods automatically return http errors.
+type ErrorResponse interface {
+	// Err returns the error set on the response.
+	//
+	// This method is named "Err" rather than "Error" to avoid confusion with
+	// the "Error" function that is specified in the error interface.
+	Err() error
+
+	// SetError sets an error to be returned in the response.
+	SetError(err error)
+}
+
+// ErrorMapper is a function used by the httpbp.Encode functions to map the
+// errors provided by responses that implement the ErrorResponse interface to
+// HTTP errors.
+type ErrorMapper func(err error) error
+
+// DefaultErrorMapper is the ErrorMapper used by the httpbp.Encode functions
+// that do not take an ErrorMapper as an argument.
+func DefaultErrorMapper(err error) error {
+	return err
+}
+
+type responseEncoder func(w http.ResponseWriter, r interface{}) error
+
+type encodeArgs struct {
+	contentType string
+	encoder     responseEncoder
+	errorMapper *ErrorMapper
+}
+
+func encodeResponse(w http.ResponseWriter, response interface{}, args encodeArgs) error {
```

Comment:
🔕 for unexported functions using a struct as arg is unnecessary (but I'm not against it either)

Reply:
Ya, I did it _more_ to keep the argument list smaller

Reply:
yea it's fine to keep it as is.